components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -3672,6 +3833,116 @@ private void updateUserClaims(User user, Map<String, String> oldClaimList,
+        Map<String, String> oldClaimListExcludingMultiValuedClaims = new HashMap<>(oldClaimList);
+        oldClaimListExcludingMultiValuedClaims.keySet().removeAll(allSimpleMultiValuedClaimsList.keySet());
+
+        Map<String, String> newClaimListExcludingMultiValuedClaims = new HashMap<>(newClaimList);
+        newClaimListExcludingMultiValuedClaims.keySet().removeAll(allSimpleMultiValuedClaimsList.keySet());
+
+        Map<String, String> userClaimsToBeAdded = new HashMap<>(newClaimListExcludingMultiValuedClaims);
+        Map<String, String> userClaimsToBeDeleted = new HashMap<>(oldClaimListExcludingMultiValuedClaims);
+        Map<String, String> userClaimsToBeModified = new HashMap<>();
+
+        // Get all the old claims, which are not available in the new claims.
+        userClaimsToBeDeleted.keySet().removeAll(newClaimListExcludingMultiValuedClaims.keySet());
+
+        // Get all the new claims, which are not available in the existing claims.
+        userClaimsToBeAdded.keySet().removeAll(oldClaimListExcludingMultiValuedClaims.keySet());
+
+        // Get all new claims, which are only modifying the value of an existing claim.
+        for (Map.Entry<String, String> eachNewClaim : newClaimListExcludingMultiValuedClaims.entrySet()) {
+            if (oldClaimListExcludingMultiValuedClaims.containsKey(eachNewClaim.getKey()) &&
+                    !oldClaimListExcludingMultiValuedClaims.get(eachNewClaim.getKey())
+                            .equals(eachNewClaim.getValue())) {
+                userClaimsToBeModified.put(eachNewClaim.getKey(), eachNewClaim.getValue());
+            }
+        }
+
+        // Remove user claims.
+        for (Map.Entry<String, String> entry : userClaimsToBeDeleted.entrySet()) {
+            if (!isImmutableClaim(entry.getKey())) {
+                carbonUM.deleteUserClaimValueWithID(user.getId(), entry.getKey(), null);
+            }
+        }
+
+        // Update user claims.
+        userClaimsToBeModified.putAll(userClaimsToBeAdded);
+        if (MapUtils.isEmpty(simpleMultiValuedClaimsToBeAdded) &&
+                MapUtils.isEmpty(simpleMultiValuedClaimsToBeRemoved)) {
+            // If no multi-valued attribute is modified.
+            carbonUM.setUserClaimValuesWithID(user.getId(), userClaimsToBeModified, null);
+        } else {
+            carbonUM.setUserClaimValuesWithID(user.getId(), convertClaimValuesToList(oldClaimList),
+                    simpleMultiValuedClaimsToBeAdded, simpleMultiValuedClaimsToBeRemoved,
+                    convertClaimValuesToList(userClaimsToBeModified), null);
+        }
+    }
+
+    /**
+     * Convert the claim values to list of strings. Main usage is for multi-valued claims.
+     *
+     * @param claimMap Map of claim URIs against claim value as string.
+     * @return Map of claim URIs against claim value as a list of string.
+     */
+    private Map<String, List<String>> convertClaimValuesToList(Map<String, String> claimMap) {
```

Comment:
`buildMapOfClaims` WDYT? In the method description we can say what we are trying to achieve.

Reply:
I feel like `buildMapOfClaims` doesn't suite as the method signature even the description says the method behavior. The reason : the parameter which passes into this method also a claim map.